airflow/contrib/sensors/emr_job_flow_sensor.py

```diff
@@ -16,48 +16,96 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+
 from airflow.contrib.hooks.emr_hook import EmrHook
 from airflow.contrib.sensors.emr_base_sensor import EmrBaseSensor
 from airflow.utils.decorators import apply_defaults
 
 
 class EmrJobFlowSensor(EmrBaseSensor):
     """
-    Asks for the state of the JobFlow until it reaches a terminal state.
+    Asks for the state of the EMR JobFlow (Cluster) until it reaches
+    any of the target states.
     If it fails the sensor errors, failing the task.
 
+    With the default target states, sensor waits cluster to be terminated.
+    When target_states is set to ['RUNNING', 'WAITING'] sensor waits
+    until job flow to be ready (after 'STARTING' and 'BOOTSTRAPPING' states)
+
     :param job_flow_id: job_flow_id to check the state of
     :type job_flow_id: str
+
+    :param target_states: the target states, sensor waits until
+        job flow reaches any of these states
+    :type target_states: list[str]
+
+    :param failed_states: the failure states, sensor fails when
+        job flow reaches any of these states
+    :type failed_states: list[str]
     """
 
-    NON_TERMINAL_STATES = ['STARTING', 'BOOTSTRAPPING', 'RUNNING',
-                           'WAITING', 'TERMINATING']
-    FAILED_STATE = ['TERMINATED_WITH_ERRORS']
-    template_fields = ['job_flow_id']
+    template_fields = ['job_flow_id', 'target_states', 'failed_states']
     template_ext = ()
 
     @apply_defaults
     def __init__(self,
                  job_flow_id,
+                 target_states=None,
+                 failed_states=None,
                  *args,
                  **kwargs):
         super().__init__(*args, **kwargs)
         self.job_flow_id = job_flow_id
+        if target_states is None:
+            target_states = ['TERMINATED']
+        self.target_states = target_states
+        if failed_states is None:
+            failed_states = ['TERMINATED_WITH_ERRORS']
+        self.failed_states = failed_states
 
     def get_emr_response(self):
+        """
+        Make an API call with boto3 and get cluster-level details.
+
+        .. seealso::
+            https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/emr.html#EMR.Client.describe_cluster
+
+        :return: response
+        :rtype: dict
+        """
         emr = EmrHook(aws_conn_id=self.aws_conn_id).get_conn()
 
         self.log.info('Poking cluster %s', self.job_flow_id)
         return emr.describe_cluster(ClusterId=self.job_flow_id)
 
     @staticmethod
     def state_from_response(response):
```

Comment:
```suggestion
    def state_from_response(response: Dict[str, Any]) -> str:
```
It's awesome that you've added the docstrings! Type annotations will help even more.